examples/pytorch/seal_ogbl/utils.py

```diff
@@ -0,0 +1,134 @@
+import sys
+import random
+import numpy as np
+from scipy.sparse.csgraph import shortest_path
+import torch
+from dgl.sampling import global_uniform_negative_sampling
+
+
+def k_hop_subgraph(src, dst, num_hops, g, sample_ratio=1., directed=False):
+    # Extract the k-hop enclosing subgraph around link (src, dst) from g
+    nodes = [src, dst]
+    visited = set([src, dst])
+    fringe = set([src, dst])
+    for _ in range(num_hops):
+        if not directed:
+            _, fringe = g.out_edges(list(fringe))
+            fringe = fringe.tolist()
+        else:
+            _, out_neighbors = g.out_edges(list(fringe))
+            in_neighbors, _ = g.in_edges(list(fringe))
+            fringe = in_neighbors.tolist() + out_neighbors.tolist()
+        fringe = set(fringe) - visited
+        visited = visited.union(fringe)
+        
+        if sample_ratio < 1.:
+            fringe = random.sample(fringe, int(sample_ratio * len(fringe)))
+        if len(fringe) == 0:
+            break
+
+        nodes = nodes + list(fringe)
+
+    subg = g.subgraph(nodes, store_ids=True)
+
+    return subg
+
+
+def drnl_node_labeling(adj, src, dst):
+    # Double Radius Node Labeling (DRNL).
+    src, dst = (dst, src) if src > dst else (src, dst)
+
+    idx = list(range(src)) + list(range(src + 1, adj.shape[0]))
+    adj_wo_src = adj[idx, :][:, idx]
+
+    idx = list(range(dst)) + list(range(dst + 1, adj.shape[0]))
+    adj_wo_dst = adj[idx, :][:, idx]
+
+    dist2src = shortest_path(adj_wo_dst, directed=False, unweighted=True, indices=src)
+    dist2src = np.insert(dist2src, dst, 0, axis=0)
+    dist2src = torch.from_numpy(dist2src)
+
+    dist2dst = shortest_path(adj_wo_src, directed=False, unweighted=True, indices=dst-1)
+    dist2dst = np.insert(dist2dst, src, 0, axis=0)
+    dist2dst = torch.from_numpy(dist2dst)
+
+    dist = dist2src + dist2dst
+    dist_over_2, dist_mod_2 = dist // 2, dist % 2
+
+    z = 1 + torch.min(dist2src, dist2dst)
+    z += dist_over_2 * (dist_over_2 + dist_mod_2 - 1)
+    z[src] = 1.
+    z[dst] = 1.
+    # shortest path may include inf values
+    z[torch.isnan(z)] = 0.
+
+    return z.to(torch.long)
+
+
+def get_pos_neg_edges(split, split_edge, g, percent=100):
+    assert 'edge' in split_edge['train']
```

Comment:
Why do we need this assertion? Perhaps you should either leave a comment for elaboration or remove this line.